Creating callUnboxedWithDispatchKey method

charliechen0401 opened the pull request:
Summary:
creating a method called "callUnboxedWithDispatchKey".

Also adding tests to make sure it works.

Test Plan: buck test mode/dev //caffe2:ATen-core-test

Differential Revision: D19402815

charliechen0401 commented on aten/src/ATen/core/op_registration/op_registration_test.cpp:
```diff
@@ -774,6 +774,32 @@ void autograd_kernel(Tensor a) {
   called_autograd = true;
 }
 
+TEST(OperatorRegistrationTest, whenRegisteringCPUTensorType_thenCanCallUnboxedWithCPUTensorIdDispatchKey) {
```
@smessmer , let me know if this is the ideal way to add test? (is this the correct place to add unit tests? also if the way to test make sense.

smessmer replied:
This is unrelated to autograd, so I'd rather add the test somewhere further up in the file where we test that we can dispatch to different backends. Somewhere close to whenRegisteringMultipleKernelsInSameOpCallAndCalling_thenCallsCorrectKernel

Commit:
Creating callUnboxedWithDispatchKey method (#32198)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32198

creating a method called "callUnboxedWithDispatchKey".

Also adding tests to make sure it works.

Test Plan: buck test mode/dev //caffe2:ATen-core-test

Differential Revision: D19402815

fbshipit-source-id: 3de7aecd53c602e125cb8b057414d4f07d47e00b

smessmer commented on aten/src/ATen/core/op_registration/op_registration_test.cpp:
```diff
@@ -505,6 +505,25 @@ TEST(OperatorRegistrationTest, givenMultipleCatchallKernels_whenNewerAndThenOlde
+  ASSERT_TRUE(op.has_value()); // assert schema is registered
+
+  called_kernel_cpu = false;
+  callOpUnboxedWithDispatchKey<void, Tensor>(*op, c10::DispatchKey::CPUTensorId, dummyTensor(c10::DispatchKey::CPUTensorId));
```
Maybe pass a `dummyTensor(CUDATensorId)` here to make the test more meaningful and ensure that it doesn't take the dispatch key from the tensor but from the direct argument instead.

smessmer commented on aten/src/ATen/core/op_registration/op_registration_test.cpp:
```diff
@@ -505,6 +505,25 @@ TEST(OperatorRegistrationTest, givenMultipleCatchallKernels_whenNewerAndThenOlde
+
+  called_kernel_cpu = false;
+  expectThrows<c10::Error>([&] {
+    callOpUnboxedWithDispatchKey<void, Tensor>(*op, c10::DispatchKey::CUDATensorId, dummyTensor(c10::DispatchKey::CUDATensorId));
```
also here change the `dummyTensor`
